guides/docs/testing/testing_controllers.md

```diff
@@ -133,169 +222,489 @@ defmodule HelloWeb.Router do
+# test/hello_web/controllers/user_controller_test.exs line 55 (or so)
+
+test "Responds with a message indicating user not found", %{conn:  conn} do
+  response = conn
+  |> get(user_path(conn, :show, -1 ))
   |> json_response(404)
 
-  expected = %{ "error" => "User not found." }
+  expected = %{
+    "errors" => "Resource not found"
+  }
 
   assert response == expected
 end
 ```
 
 We want a HTTP code of 404 to notify the requester that this resource was not found, as well as an accompanying error message.
 
-Our controller action:
+Our controller action now needs to handle the error thrown by Ecto:
 
 ```elixir
-def show(conn, %{"id" => id}) do
-  case Repo.get(User, id) do
-    nil -> conn |> put_status(404) |> render("error.json")
-    user -> render conn, "show.json", user: user
+# lib/hello_web/controllers/user_controller.ex, line 10 or so
+
+  def show(conn, %{"id" => id}) do
+    try do
+      user = Accounts.get_user!(id)
+      render conn, "show.json", data: user
+    catch
+      :error, message ->
+        conn
+        |> put_status(:not_found)
+        |> render( HelloWeb.ErrorView, "400.json", reason: message )
+    end
   end
-end
 ```
 
-And our view:
+In the `catch` block, there are two important things going on.
+First we use the
+[`put_status/2`](https://hexdocs.pm/plug/Plug.Conn.html#put_status/2)
+function from `Plug.Conn` to set the desired error status.  The complete list of allowed codes can be found in the
+[Plug.Conn.Status
+documentation](https://hexdocs.pm/plug/Plug.Conn.Status.html), where
+we can see that `:not_found` corresponds to our desired "404" status.
 
-```elixir
-def render("error.json", _assigns) do
-  %{error: "User not found."}
+Second, we've redirected the view to the `ErrorView` module.  If you
+open that file up, you'll see a few rendering helpers for HTML,
+output, but none for JSON.  We need to add a render function for
+"400.json" as follows:
+
+```
+# lib/hello_web/views/error_view.ex
+
+defmodule HelloWeb.ErrorView do
+  use HelloWeb, :view
+
+  def render("404.html", _assigns) do
+    "Page not found"
+  end
+
+  def render("500.html", _assigns) do
+    "Internal server error"
+  end
+
+  def render("400.json", %{reason: reason}) do
+    message = case reason do
+                %Ecto.NoResultsError{} -> "Resource not found"
+                _ -> "I'm afraid I can't do that, Dave"
+              end
+    %{errors: message}
+  end
+
+  # In case no render clause matches or no
+  # template is found, let's render it as 500
+  def template_not_found(_template, assigns) do
+    render "500.html", assigns
+  end
 end
 ```
 
+In the "400.json" render function, we only expect the `%Ecto.NoResultsError{}`
+so far, so we only set that message.  As other errors crop up in our
+TDD development, we can add additional error messages.
+
 With those implemented, our tests pass.
 
-The rest of the controller is left to you to implement as practice. Happy testing!
+### Two down, six (or more) to go
```

Comment:
I don't think we nee d a new heading here. We can say:

 > The rest of the controller is left for you to implement as practice. If you are not sure where to begin, it is worth using the Phoenix JSON generator and seeing what tests are automatically generated for you.

Reply:
Agreed.  I interpret your intent to mean I should delete the sample test block as well, correct?  No one is really going to do it this way, they're going to use the phx.gen.json generator and edit that.

Reply:
Your interpretation is correct.

Reply:
fixed